### ISSUE TITLE
Broken vulkan surface type VK_KHR_xcb_surface with nvidia_x11 version 390.48

### ISSUE THREAD

**Issue by danme** [NixOS/nixpkgs#39149]
## Issue description

Since I upgraded from NixOS 17.09 to NixOS 18.03, I can't get the vulkan demos running.
The vulkan-loader.demos `cube` executable exits just after start with a Segmentation Fault.

A look into `vulkaninfo` shows that the Presentable Surfaces (xcb) do not have any Formats (count = 0):
```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.61

INFO: [loader] Code 0 : Meta-layer VK_LAYER_LUNARG_standard_validation all 0 component layers appear to be valid.
INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver/share/vulkan/icd.d/nvidia.json, version "1.0.0"
INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver-32/share/vulkan/icd.d/nvidia.json, version "1.0.0"

...

Presentable Surfaces:
=====================
GPU id       : 0 (NVS 510)
Surface type : VK_KHR_xcb_surface
Formats:                count = 0
Present Modes:          count = 3
        FIFO_KHR
        FIFO_RELAXED_KHR
        IMMEDIATE_KHR
```

### Good case NixOS 17.09

In `Linux 4.9.86, NixOS, 17.09pre117521.e984f9e48e1 (Hummingbird)` width nvidia "384.111"
I got the expected result again:
```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.42

INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver/share/vulkan/icd.d/nvidia.json, version "1.0.0"
INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver-32/share/vulkan/icd.d/nvidia.json, version "1.0.0"

...

Presentable Surfaces:
=====================
GPU id       : 0 (NVS 510)
Surface type : VK_KHR_xcb_surface
Formats:                count = 2
        B8G8R8A8_UNORM
        B8G8R8A8_SRGB
Present Modes:          count = 3
        FIFO_KHR
        FIFO_RELAXED_KHR
        IMMEDIATE_KHR

GPU id       : 1 (NVS 510)
Surface type : VK_KHR_xcb_surface
Formats:                count = 2
        B8G8R8A8_UNORM
        B8G8R8A8_SRGB
Present Modes:          count = 3
        FIFO_KHR
        FIFO_RELAXED_KHR
        IMMEDIATE_KHR

```

### Testing NixOS 17.09 with nvidia 390.48

Same behavior as for NixOS 18.03: Formats count = 0.

### Workaround VK_KHR_xlib_surface

In NixOS 18.03 with nvidia 390.48.

Further investigations show that this is a problem of the VK_KHR_xcb_surface but VK_KHR_xlib_surface is not affected.
Because when I patch the pkgs/development/libraries/vulkan-loader/default.nix to use Xlib:
```
  cmakeFlags = [
    "-DBUILD_WSI_MIR_SUPPORT=OFF"
    "-DBUILD_WSI_XCB_SUPPORT=OFF"
    "-DDEMOS_WSI_SELECTION=XLIB"
    "-DFALLBACK_DATA_DIRS=${libGL.driverLink}/share:/usr/local/share:/usr/share"
  ];
```
I get valid results from vulkaninfo:
```
===========
VULKAN INFO
===========

Vulkan API Version: 1.0.61

INFO: [loader] Code 0 : Meta-layer VK_LAYER_LUNARG_standard_validation all 0 component layers appear to be valid.
INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver/share/vulkan/icd.d/nvidia.json, version "1.0.0"
INFO: [loader] Code 0 : Found ICD manifest file /run/opengl-driver-32/share/vulkan/icd.d/nvidia.json, version "1.0.0"

...

Presentable Surfaces:
=====================
GPU id       : 0 (NVS 510)
Surface type : VK_KHR_xlib_surface
Formats:                count = 2
        B8G8R8A8_UNORM
        B8G8R8A8_SRGB
Present Modes:          count = 3
        FIFO_KHR
        FIFO_RELAXED_KHR
        IMMEDIATE_KHR

```

### Testing vulkan-loader 1.1.70 (borrowed from nixos-unstable)

On NixOS 18.03 with nvidia 390.48: Formats count = 0.

But in NixOS 17.09 with nvidia 384.111 it works: Format count = 2.

### Conclusion

I assume that this is a problem of the nvidia driver 390.48 in combination with a XCB surface, Upstream bug?

### Steps to reproduce

Test cube demo (crashes with a Segmentation fault):
1. `nix-shell -p vulkan-loader.demos`
2. `cube`

Inspecting vulkaninfo:
1. `nix-shell -p vulkan-loader`
2. `vulkaninfo | less`

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.34, NixOS, 18.03pre132021.c0c5571ec1a (Impala)`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.0pre6137_e3cdcf89`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/vas/nixpkgs`

**Comment by danme** [comment]
Reported the bug at Nvidia as well: https://developer.nvidia.com/nvidia_bug/2108502

**Comment by Rotaerk** [comment]
Hmm, I just ran across this issue.  I'm seeing it as well, but I also reported it to NVIDIA here:  https://devtalk.nvidia.com/default/topic/1036182/linux/vulkan-error-with-nvidia-linux-driver-387-12-and-later/.  (Not really clear on the difference between developer.nvidia.com and devtalk.nvidia.com, but I can't see your link.  Says membership is required.)

I didn't initially notice vulkaninfo showing 0 formats available, but indeed it does.  As my post indicates, I first noticed a failure during the call to vkGetPhysicalDeviceSurfaceFormatsKHR.

**Comment by damienleone** [comment]
@Rotaerk Your post on our forums brought me here.  I had not noticed this bug report before.

I took a look and found that the issue was due to our Vulkan ICD not being able to dlopen() libxcb-glx.so.0 when initializing the XCB path.  We first dlopen() libxcb.so.1, which succeeds, then we fail to dlopen() libxcb-glx.so.0.

I am not very familiar with NixOS but from what I can tell both libxcb.so.1 and libxcb-glx.so.0 are available in the same directory (./nix/store/sjr3r24v00n4lk6aamnfgp868kj75n06-libxcb-1.12/lib)

**Comment by damienleone** [comment]
Forgot to mention that the error reported by dlerror() is:

> libxcb-glx.so.0: cannot open shared object file: No such file or directory

**Comment by Rotaerk** [comment]
Thanks @damienleone.  I think this is a NixOS-specific issue; I asked someone on a different distro to check `vulkaninfo` after having upgraded to driver version 390.67, and it showed a non-zero number of available formats.  I guess something about NixOS is making libxcb-glx.so.0 unavailable to the driver.

**Comment by ck3d** [comment]
An ugly but working workaround is putting libxcb to LD_LIBRARY_PATH.

**Comment by ck3d** [comment]
Intel GPU driver (and NVidia 384.x) do not try to load libxcb-glx.
```
 Presentable Surfaces:
=====================
GPU id       : 0 (Intel(R) HD Graphics 630 (Kaby Lake GT2))
Surface type : VK_KHR_xcb_surface
Formats:                count = 2
        B8G8R8A8_SRGB
        B8G8R8A8_UNORM
Present Modes:          count = 3
        IMMEDIATE_KHR
        MAILBOX_KHR
        FIFO_KHR
```
So, why does NVidia 390.x need to load libxcb-glx?

**Comment by damienleone** [comment]
For proper XCB support.  We use xcb_glx_query_version and xcb_glx_query_version_reply from that library.

**Comment by Infinisil** [comment]
Fixed by #46555 
Thanks :D
